rs/consensus/src/consensus/validator.rs

```diff
@@ -992,6 +1080,35 @@ impl Validator {
                 continue;
             }
 
+            // Disqualify rank if equivocation was found. If there already
+            // exists a validated block of the same rank as the current
+            // proposal, we must generate an equivocation proof.
+            if let Some(existing_metadata) =
+                valid_qualified_ranks.get(proposal.height(), proposal.rank())
+            {
```

Comment:
This is more of a general question: would it simplify the code if we disqualified signers instead of ranks? Below we anyway assume that the signers for both metadata must be equal.

Reply:
I tried that at first, but we need to query the lowest rank for a given height a few lines above, so collecting ranks in the first place seemed easiest.